File: Hard/Longest Path in a Directed Acyclic Graph/longest-path-in-a-directed-acyclic-graph.cpp
class Solution {
public:

    void dfs(int node,
             vector<vector<pair<int, int>>> &adj,
             vector<int> &vis,
             stack<int> &st) {

        vis[node] = 1;

        for (auto &it : adj[node]) {

            int v = it.first;

            if (!vis[v]) {
                dfs(v, adj, vis, st);
            }
        }

        st.push(node);
    }

    vector<int> maxDistance(int V, int src, vector<vector<int>> &edges) {

        vector<vector<pair<int, int>>> adj(V);
        stack<int> st;

        // Build graph with negative weights
        for (auto &edge : edges) {

            int u = edge[0];
            int v = edge[1];
            int w = edge[2];

            adj[u].push_back({v, -w});
        }

        vector<int> dist(V, INT_MAX);
        dist[src] = 0;

        vector<int> vis(V, 0);

        // Topological sort
        for (int i = 0; i < V; i++) {

            if (!vis[i]) {
                dfs(i, adj, vis, st);
            }
        }

        // Shortest path on DAG
        while (!st.empty()) {

            int cur = st.top();
            st.pop();

            // Skip nodes unreachable from src
            if (dist[cur] == INT_MAX)
                continue;

            for (auto &it : adj[cur]) {

                int v = it.first;
                int w = it.second;

                if (dist[cur] + w < dist[v]) {
                    dist[v] = dist[cur] + w;
                }
            }
        }

        // Convert back to maximum distances
        for (int i = 0; i < V; i++) {

            if (dist[i] != INT_MAX) {
                dist[i] *= -1;
            }
            else {
                dist[i] = INT_MIN;
            }
        }

        return dist;
    }
};

File: Hard/Longest Path in a Directed Acyclic Graph/longest-path-in-a-directed-acyclic-graph.py
class Solution:
    def maxDistance(self, V, src, edges):
        # code here
        adj = [[] for _ in range(V)]
        st = []
        for u,v,w in edges:
            adj[u].append([v , -1*w])
        
        dist = [float('INF')] * V
        
        dist[src] = 0
        
        vis = [0]*V
        
        
        def dfs(node):
            vis[node] = 1
            
            for i in adj[node]:
                v = i[0]
                if not vis[v]:
                    dfs(v)
            
            st.append(node)
            
            
        for i in range(V):
            if not vis[i]:
                dfs(i)
        
        
        while (len(st) > 0):
            cur = st.pop()
            
            for v,w in adj[cur]:
                if dist[cur]+w < dist[v]:
                    dist[v] = dist[cur] + w
        
        for i in range(V):
            if dist[i] != float('INF'):
                dist[i] *= -1
            else:
                dist[i] = (-2**31)
        
        return dist
        
        
            

